File: dogfight/src/entities/water.ts
import { FacingDirection } from "../constants";
import { Entity, EntityType } from "../entity";
import { Cache, CacheEntry } from "../network/cache";
import { RectangleBody, Rectangle } from "../physics/rectangle";
import { GameWorld } from "../world/world";
import { SolidEntity } from "./SolidEntity";

export class Water extends SolidEntity {
  public type = EntityType.Water;

  public x: number;
  public y: number;
  public width: number;
  public direction: FacingDirection;

  public constructor(id: number, world: GameWorld, cache: Cache) {
    super(id, world, -1);
    this.setData(cache, {
      x: 0,
      y: 0,
      width: 30000,
      direction: FacingDirection.Right
    });
  }

  public getCollisionBounds(): Rectangle {
    return new Rectangle(this.x, this.y + 5 - 50, this.width, 100);
  }

  public getState(): CacheEntry {
    return {
      type: this.type,
      x: this.x,
      y: this.y,
      width: this.width,
      direction: this.direction
    };
  }

}

export function getWaterRect(
  x: number,
  y: number,
  width: number
): RectangleBody {
  return {
    direction: 0,
    width,
    height: 1000,
    center: {
      x,
      y: y - 510
    }
  };
}

File: dogfight/src/entities/water.rs
use dogfight_macros::Networked;
use image::RgbaImage;

use crate::{
    collision::{BoundingBox, SolidEntity},
    network::{property::*, EntityProperties, NetworkedEntity},
};

use super::{
    entity::Entity,
    types::{EntityType, Facing, Terrain},
};

#[derive(Networked)]
pub struct Water {
    terrain: Property<Terrain>,
    facing: Property<Facing>,
    width: Property<i16>,
    client_x: Property<i16>,
    client_y: Property<i16>,
}

impl Water {
    pub fn new(terrain: Terrain, facing: Facing, width: i16, x: i16, y: i16) -> Self {
        Self {
            terrain: Property::new(terrain),
            facing: Property::new(facing),
            width: Property::new(width),
            client_x: Property::new(x),
            client_y: Property::new(y),
        }
    }
}

impl Entity for Water {
    fn get_type(&self) -> EntityType {
        EntityType::Water
    }
}

impl SolidEntity for Water {
    fn get_collision_bounds(&self) -> BoundingBox {
        BoundingBox {
            x: *self.client_x.get(),
            y: *self.client_y.get() + 5,
            width: *self.width.get(),
            height: 100,
        }
    }

    fn get_collision_image(&self) -> Option<&RgbaImage> {
        None
    }
}
